Fabric.Terminology.SqlServer/Persistence/UnitOfWork/DtoDataReader.cs

```diff
@@ -0,0 +1,139 @@
+﻿namespace Fabric.Terminology.SqlServer.Persistence.UnitOfWork
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Data.Common;
+    using System.Linq;
+    using System.Reflection;
+
+    internal class DtoDataReader<T> : DbDataReader
+        where T : class
+    {
+        private readonly IReadOnlyCollection<T> dtos;
+
+        private readonly IEnumerator<T> enumerator;
+
+        private readonly List<PropertyInfo> properties = new List<PropertyInfo>();
+
+        private readonly Dictionary<string, int> nameLookup = new Dictionary<string, int>();
+
+        private DtoDataReader(IEnumerable<T> dtos)
+        {
+            this.dtos = dtos.ToList();
+            this.enumerator = this.dtos.GetEnumerator();
+
+            this.properties.AddRange(
+                typeof(T).GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.DeclaredOnly));
+
+            for (var i = 0; i < this.properties.Count; i++)
+            {
+                var name = this.properties[i].Name.Equals("CodeSystemGuid", StringComparison.OrdinalIgnoreCase)
+                               ? "CodeSystemGUID"
+                               : this.properties[i].Name;
+
+                this.nameLookup[name] = i;
+            }
+        }
+
+        public override int FieldCount => this.properties.Count;
+
+        public override int RecordsAffected => this.dtos.Count;
+
+        public override bool HasRows => this.dtos != null && this.dtos.Any();
+
+        public override bool IsClosed => this.enumerator == null;
+
+        public override object this[int ordinal] => this.GetValue(ordinal);
+
+        public override object this[string name] => this.GetValue(this.GetOrdinal(name));
+
+        public static DtoDataReader<T> Create(IEnumerable<T> dtos) => new DtoDataReader<T>(dtos);
+
+        public override bool Read() => this.enumerator.MoveNext();
+
+        public override IEnumerator GetEnumerator() => this.enumerator;
+
+        public override int GetOrdinal(string name)
+        {
+          return this.nameLookup.ContainsKey(name) ? this.nameLookup[name] : -1;
+        } 
+
+        public override object GetValue(int ordinal) => this.properties[ordinal].GetValue(this.enumerator.Current, null);
+
+        public override int GetValues(object[] values)
+        {
+            var getValues = Math.Max(this.FieldCount, values.Length);
```

Comment:
From what I can tell, this will only have the effect of causing line 70 to throw an exception. Why not fail fast with an explicit message?

```
if(values.Length != this.FieldCount)
{
    throw new InvalidOperationException("The provided values had length of ... which does not match the length of this reader's fields (...)");
}
for (var i = 0; i < values.Length; i++)
...
```